framework/include/constraints/NodeElemConstraint.h

```diff
@@ -0,0 +1,280 @@
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef NodeElemConstraint_H
+#define NodeElemConstraint_H
+
+// MOOSE includes
+#include "Constraint.h"
+#include "NeighborCoupleableMooseVariableDependencyIntermediateInterface.h"
+
+// Forward Declarations
+class NodeElemConstraint;
+
+// libMesh forward declarations
+namespace libMesh
+{
+template <typename T>
+class SparseMatrix;
+}
+
+template <>
+InputParameters validParams<NodeElemConstraint>();
+
+/**
+ * A NodeElemConstraint is used when you need to create constraints between
+ * two overlapping blocks. It works by allowing you to modify the residual
+ * and jacobian entries on "this" side (the node side, also referred to as
+ * the slave side) and the "other" side (the elem side, also referred to as
+ * the master side)
+ */
+class NodeElemConstraint : public Constraint,
+                           public NeighborCoupleableMooseVariableDependencyIntermediateInterface,
+                           public NeighborMooseVariableInterface<Real>
+{
+public:
+  NodeElemConstraint(const InputParameters & parameters);
+  virtual ~NodeElemConstraint();
+
+  /**
+   * Compute the value the slave node should have at the beginning of a timestep.
+   */
+  void computeSlaveValue(NumericVector<Number> & current_solution);
+
+  /**
+   * Computes the residual Nodal residual.
+   */
+  virtual void computeResidual();
+
+  /**
+   * Computes the jacobian for the current element.
+   */
+  virtual void computeJacobian();
+
+  /**
+   * Computes d-residual / d-jvar...
+   */
+  virtual void computeOffDiagJacobian(unsigned int jvar);
+
+  /**
+   * Gets the indices for all dofs connected to the constraint
+   */
+  virtual void getConnectedDofIndices(unsigned int var_num);
+
+  /**
+   * Whether or not this constraint should be applied.
+   *
+   * Get's called once per slave node per variable.
+   */
+  virtual bool shouldApply() { return true; }
+
+  /**
+   * Compute reaction force of the constraint depending on the formulation
+   */
+  virtual void computeReactionForce(){};
+
+  /**
+   * Whether or not the slave's residual should be overwritten.
+   *
+   * When this returns true the slave's residual as computed by the constraint will _replace_
+   * the residual previously at that node for that variable.
+   */
+  virtual bool overwriteSlaveResidual();
+
+  /**
+   * Whether or not the slave's Jacobian row should be overwritten.
+   *
+   * When this returns true the slave's Jacobian row as computed by the constraint will _replace_
+   * the residual previously at that node for that variable.
+   */
+  virtual bool overwriteSlaveJacobian() { return overwriteSlaveResidual(); };
+
+  /**
+   * The variable on the Master side of the domain.
+   */
+  virtual MooseVariable & masterVariable() { return _master_var; }
+
+  // TODO: Make this protected or add an accessor
+  // Do the same for all the other public members
+  SparseMatrix<Number> * _jacobian;
```

Comment:
This must have been copied from another constraint. Go ahead and do what is in the TODO comment since we're starting fresh here.